### PR TITLE
Lighting Should Only hold Vec<Entity> instead of TypeId<Vec<Entity>>

### PR DESCRIPTION
**Pull request by re0312**
# Objective
- After #13894, I noticed the performance of `many_lights `dropped from 120+ to 60+. I reviewed the PR but couldn't identify any mistakes. After profiling, I discovered that `Hashmap::Clone `was very slow when its not empty, causing `extract_light` to increase from 3ms to 8ms.
- Lighting only checks visibility for 3D Meshes. We don't need to maintain a TypeIdMap for this, as it not only impacts performance negatively but also reduces ergonomics.

## Solution

- use VisibleMeshEntities for lighint visibility checking.

## Performance
cargo run --release --example many_lights  --features bevy/trace_tracy 
name="bevy_pbr::light::check_point_light_mesh_visibility"}
image

system{name="bevy_pbr::render::light::extract_lights"}
image

## Migration Guide

> now `SpotLightBundle` , `CascadesVisibleEntities  `and `CubemapVisibleEntities `use VisibleMeshEntities instead of `VisibleEntities`

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 1]
```diff
@@ -230,6 +230,13 @@ impl VisibleEntities {
     }
 }
 
+#[derive(Component, Clone, Debug, Default, Reflect, Deref, DerefMut)]
+#[reflect(Component)]
+pub struct VisibleMeshEntities {
```
Comment (mnmaita) [inline crates/bevy_render/src/view/visibility/mod.rs:235]: A few doc comments would be great to be able to differentiate this from `VisibleEntities`

File: crates/bevy_pbr/src/bundle.rs [hunk 2]
```diff
@@ -45,27 +46,37 @@ impl<M: Material> Default for MaterialMeshBundle<M> {
     }
 }
 
+/// Collection of mesh entities visible for 3D lighting.
+/// This component contains all mesh entities visible from the current light view.
+/// The collection is updated automatically by  [`SimulationLightSystems::CheckLightVisibility`].
```
Comment (pcwalton) [inline crates/bevy_pbr/src/bundle.rs:51]: nit: extra space before `[`

**Review by pcwalton — APPROVED** [review]
Seems fine. Lights could in theory affect other types of objects, but that would require a lot of refactoring, so we can specialize them to `Mesh`es for now.

File: crates/bevy_pbr/src/bundle.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_pbr/src/bundle.rs:51]:
```suggestion
/// The collection is updated automatically by [`SimulationLightSystems::CheckLightVisibility`].
```

**Pull request merged**